### ISSUE TITLE
I just implemented contact blocking....

### ISSUE THREAD

**Issue by agilob** [qTox/qTox#2909]
what should happen when you block a contact? Should notifications for a contact be completely off? Should we save and show new messages but no sound and no notification? Should contact be always visible as offline until unblocked?

**Comment by tux3** [comment]
The contact should be temporarily removed from the friendlist so that they can't see you as online or contact you in any way while blocked.

**Comment by agilob** [comment]
Where is code responsible for sending status to other contacts?

**Comment by tux3** [comment]
In toxcore.

**Comment by TheNain38** [comment]
Normally you should remove contact from friendlist like @tux3 said, and when unblocking you should readd it using

``` c
uint32_t tox_friend_add_norequest(Tox *tox, const uint8_t *public_key, TOX_ERR_FRIEND_ADD *error);
```

**Comment by zetok** [comment]
related  #1120

**Comment by LuccoJ** [comment]
If I may ask a question... would any implementation of contact blocking let the blocker stop the blocked from being able to see their IP address?
Since my understanding is that IP addresses are stored in the DHT encrypted, but once you give someone the key by adding them as friends, they can decrypt them... and that doesn't sound like it can be revoked.

**Comment by zetok** [comment]
> Since my understanding is that IP addresses are stored in the DHT encrypted

They're not encrypted.

Edit: and to answer your question - yes, blocking someone makes toxcore not connect to them, thus they wouldn't get your IP.

**Comment by agilob** [comment]
@TheNain38 thanks for that! I'll craft a PR soon.

**Comment by TheNain38** [comment]
@agilob You're welcome! This is the right way to go:
- Blocking:
  1. Add public key to blocklist
  2. Remove from friendlist
- Unblocking:
  1. Get public key from blocklist and add contact using `tox_friend_add_norequest`
  2. Remove public key from blocklist

**Comment by ovalseven8** [comment]
It should also be possible to block a contact you don't have in your contact list. For example to prevent friend request spamming.

**Comment by agilob** [comment]
> For example to prevent friend request spamming.

That's different issue.

**Comment by LuccoJ** [comment]
@zetok if they're not encrypted, then why does the Tox wiki claim that other people (who aren't your friends) can't see your IP?
It says: 

> However, a user cannot uncover another user's IP address using only a Tox ID to find the user in the DHT, the IP address will only be discernible when the users become “friends.”

at https://wiki.tox.chat/users/faq#does_tox_leak_my_ip_address and citing https://gist.github.com/irungentoo/cb0a9b07131e95be1e64 as the source.

But if you're right and the wiki is wrong, then this is a security issue: I definitely don't want to be broadcasting my IP in plaintext to the world.
And encrypting it with a key so that only people in my contact list can unscramble it _should_ be possible, and _should_ be done.

**Comment by ovalseven8** [comment]
> For example to prevent friend request spamming.
> 
> > That's different issue.

Isn't it contact blocking?

**Comment by TheNain38** [comment]
Yes, friend request comming from a certain Public Key that is blocked shouldn't appear. I think it should also be done, so contact blocking would be complete? @agilob

**Comment by LuccoJ** [comment]
It's similar from an end-user point of view, but probably not from the developer's point of view. We do already have NOSPAM to deal with this, however...

**Comment by TheNain38** [comment]
@LuccoJ I think he meant, that even when knowing the NoSpam, you shouldn't see friend request from the blocked contact.

**Comment by zetok** [comment]
@LuccoJ 

> @zetok if they're not encrypted, then why does the Tox wiki claim that other people (who aren't your friends) can't see your IP?

Err, wiki isn't wrong, wording is slightly inaccurate though, and you should read the source it links to. Note though that it doesn't claim that your IP can't be found at all, it just says that if someone has only your ID, they can't associate it with your IP (unless you accepted their FR). Even that isn't _really_ true in all threat models, so.

> (…) this is a security issue: I definitely don't want to be broadcasting my IP in plaintext to the world.

Err, that's how p2p works. And no, it's not a security issue.

Good news though is that it's not in "plaintext" - DHT communication is also encrypted. ;)
"Bad" news is that it still doesn't matter if there's a need to scrap every IP from DHT over time.

Eh.
I wonder, given that irungentoo was supposed to write docs, whether there isn't some description somewhere of how it actually works and how it differs from "plaintext" DHTs, so that you could read more on that.

> And encrypting it with a key so that only people in my contact list can unscramble it should be possible, and should be done.

You might want to inquire about that on toxcore repo.

**Comment by ProMcTagonist** [comment]
What happened?

**Comment by agilob** [comment]
I stopped working on qTox after greyhatter and irungentoo closed some major issues in toxcore as won't fix. Changes made for this issue were deleted and my question above answered.
